lib/io/i2c.cpp

```diff
@@ -0,0 +1,85 @@
+#include "i2c.hpp"
+
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <sys/ioctl.h>
+
+#if LINUX
+#include <linux/i2c-dev.h>
+#else
+#define I2C_SLAVE 0x0703  // To specify that we are making I2C transactions
+#endif
+
+namespace hyped::io {
+
+I2c::I2c(const uint8_t bus_address) : sensor_address_(0)
+{
+  char path[13];  // up to "/dev/i2c-2"
+  sprintf(path, "/dev/i2c-%d", bus_address);
+  file_descriptor_ = open(path, O_RDWR, 0);
+  if (file_descriptor_ < 0) { /* log "Could not open i2c device" */ };
+}
+
+I2c::~I2c()
+{
+  close(file_descriptor_);
+}
+
+void I2c::setSensorAddress(const uint8_t device_address)
+{
+  if (file_descriptor_ < 0) {
+    /* log "Could not find i2c device while setting sensor address"*/;
+    return;
+  }
+  sensor_address_        = device_address;
+  const int return_value = ioctl(file_descriptor_, I2C_SLAVE, device_address);
+  if (return_value < 0) {
+    /*log "Could not set sensor address" */;
+    return;
+  }
+}
+
+std::optional<uint8_t> I2c::readByte(const uint8_t device_address, const uint8_t register_address)
```

Comment:
Do we expect to mostly read single bytes? Doing repeated reads like that may end up being both inefficient and ugly from the callees perspective.

Reply:
All of the i2c hardware only uses 8 bit registers and single byte communication so we're not concerned about multiple bytes read (atleast yet)

Reply:
Also, though it's possible that we need to read from two registers that are back to back in memory (for example data_high and data_low being at 0x06 and 0x07) and having  `readBytes` would help, but I think having one function read from both makes it harder to understand what the code is doing. (Plus I don't think we are going to need to read more than 2 registers at one time)